samcli/lib/cookiecutter/interactive_flow.py

```diff
@@ -0,0 +1,92 @@
+"""A flow of questions to be asked to the user in an interactive way."""
+from typing import Any, Dict, Optional
+import click
+from .question import Question
+
+
+class InteractiveFlow:
+    """
+    This class depends on the click module to prompt the user for answers to the questions of the flow.
+
+    Attributes
+    ----------
+    _questions: Dict[str, Question]
+        The questions to prompt the user for its answers
+    _first_question_key: str
+        The entry-point of the flow, it tells which question to ask the user first
+    """
+
+    def __init__(self, questions: Dict[str, Question], first_question_key: str):
+        self._questions: Dict[str, Question] = questions
+        self._first_question_key: str = first_question_key
+        self._current_question: Optional[Question] = None
+
+    @staticmethod
+    def _ask_a_question(question: Question) -> Any:
+        return click.prompt(text=question.text, default=question.default_answer)
+
+    @staticmethod
+    def _ask_a_yes_no_question(question: Question) -> bool:
+        return click.confirm(text=question.text, default=question.default_answer)
+
+    @staticmethod
+    def _ask_a_multiple_choice_question(question: Question) -> Any:
+        click.echo(question.text)
+        for index, option in enumerate(question.options):
+            click.echo(f"\t{index + 1} - {option}")
+        choice = click.prompt(
+            text="Choice",
+            default=question.default_answer,
+            show_choices=False,
+            type=click.Choice(question.get_choices_indexes(base=1)),
+        )
+        return question.get_option(int(choice) - 1)
+
+    def advance_to_next_question(self, current_answer: Optional[Any] = None) -> Optional[Question]:
+        """
+        Based on the answer of the current question, what is the next question to ask.
+
+        Parameters
+        ----------
+        current_answer: str
+            User's answer of the current question. This can be None of this is the first question or an optional
+            question and the user didn't provide an answer to
+
+        Returns: The next question defined in current question's next-question map or the first question if this
+                is the start of the flow.
+        """
+        if not self._current_question:
+            self._current_question = self._questions.get(self._first_question_key)
+        else:
+            next_question_key = self._current_question.get_next_question_key(current_answer)
+            self._current_question = self._questions.get(next_question_key) if next_question_key else None
+        return self._current_question
+
+    def run(self, context: Dict) -> Dict:
```

Comment:
is there any pre-processing we need to do before? Such as

* if a user has passed in prompts via cli flags, do those questions need to be skipped?